### ISSUE TITLE
Transform Bower GitHub Artifact Names to Bower Artifact Names

### ISSUE THREAD

**Issue by jamesward** [webjars/webjars#1452]
When deploying Bower WebJars and when creating dependency definitions try to transform GitHub repo-based artifact names to Bower artifact names.  See #1440

**Comment by jamesward** [comment]
I've been thinking about this and I'm leaning towards not doing this, even though I already did the work: https://github.com/webjars/webjars/compare/resolve-artifact-ids?expand=1

The reason is that I don't like changing the metadata from what the author specified.  If the author specified a dependency on a Git repo then we should reflect that in our metadata.

So I need to noodle this a bit more.

**Comment by bartolom** [comment]
as describe in #1440 I need a similar solution, but I'm solving it slightly differently: I do the translation on runtime (startup-time) in webjars-locator-core.

My source code is still kind of an internal proof of concept, but I hope to mature it in the next weeks a bit (plenty of unnecessary Logging etc. offline with service worker, ...). But because you are reasoning about #1452 at the moment, it might interest you. 

My scope is compared to the whole webjars.org project very limited as I'm only focused in the Polymer stuff at the moment, so It might be way easier for me to tweak stuff then for you.

The implementation for the is actually quite simple. If you like to have an early view you can find it here:
https://github.com/bartolom/webjars-cleaner-polymer/blob/master/src/main/java/eu/codesociety/webjars/cleaner/polymer/PomValueExtractor.java

Because the possibly expensive classpath scan is done by WebJarAssetLocator at startup-time anyways I use it to collect the pom.xml files and then be able to map webjar-qualified-github-id to github-project names.

F.i.: from `platinum-bluetooth` to `github-com-PolymerElements-platinum-bluetooth`

A use case can look like this:
https://github.com/bartolom/webjars-cleaner-polymer-sample/blob/master/src/main/java/eu/codesociety/webjars/cleaner/polymer/sample/AppConfigMvc.java#L41

The necessary change to WebJarAssetLocator is here:
https://github.com/bartolom/webjars-locator-core/commit/efd50fe073816dec4107fe25dd23359c6013eacf


I have extended `org.webjars.WebJarAssetLocator.WebJarAssetLocator(SortedMap<String, String>, WebJarFunction<String, String>)` with kind of a functional interface, which intercepts each call to gettFullPath(). 

But please be aware I'm no expert on Bower. So I simply assume that the github project name is the bower-id. This works well for the Polymer dependencies, but I'm not sure if this is always true. But one could parse the bower.json file at startup-time and use the official value from there.

Generally speaking this approach works for our use case. The Spring Boot 1.4. app is starting and I can use Polymer without any NPM/Bower stuff. Our continuous integration pipeline and servers work best with "Java-Only" so this is my main objective.

I'm trying to evolve the webjars-cleaner-polymer a bit more, and maybe write a blog post or similar at the end of the summer.

**Comment by bartolom** [comment]
btw: I'm still struggling a bit with the mixed case stuff. So I need to add excludes in my Gradle file.
https://github.com/bartolom/webjars-cleaner-polymer-sample/blob/master/build.gradle#L39

And the way to find and fix the wrong dependencies is quite tedious but possible. 

Working with the "qualified/full-github-id" in Gradle/Bower feels ok for my as a Java Person.

**Comment by bartolom** [comment]
Just a performance idea: As Webjars is actually creating the pom.xml and the pom.properties it might add two additional lines to pom.properties with the bower-id like:
```

#Generated by WebJar Sync
version=2.1.1
groupId=org.webjars.bower
artifactId=github-com-PolymerElements-platinum-bluetooth

bowerId=platinum-bluetooth
githubName=PolymerElements/platinum-bluetooth
```

and people like my do not have the pom.xml parsing with STAX at startup

**Comment by jamesward** [comment]
Thanks @bartolom for the feedback.  Could you just parse the `bower.json` files to get the metadata?

**Comment by bartolom** [comment]
Well you are right, because the bower.json file is always at the same relative position form the pom.xml file this could work. There is a Jackson Streaming API, but I have not experience with it yet.

Well one of the things which made me switch to the pom.xml in my first approach:

bower files are written by people that care about JavaScript (which is fine of course, and they achieve great things with it, which I would love to benefit from as well), but they have quite some variation:

Take for instance:
- http://cdn.jsdelivr.net/webjars/org.webjars.bower/polymer-polymer/1.6.1/bower.json
- http://cdn.jsdelivr.net/webjars/org.webjars.bower/github-com-PolymerElements-platinum-bluetooth/2.1.1/bower.json

They look quite different, there seems to be no standard place to add the github repository which is maybe evenoptional (homepage vs url)

pom.xm files are generated by a piece of software (webjars) which caters for the need of developers on the JVM, so I kind of... trust it a bit more, and because they are autogenerated they seem to be more consistent.

Also there are github project-accounts that contain dashes and could contain github-projects that contain dashes as well:

For instance
https://github.com/spring-projects/spring-framework/
I know this is not a webjar, but the "dash" is important for Polymer/WebComponents so this is not unlikly to happen.

The "qualified" webjar name: `github-com-spring-project-spring-framework` would be anbiguous

**Comment by jamesward** [comment]
We auto-generate the `pom.xml` metadata from the `bower.json` file.  But sometimes we have to go through a bunch of pain to figure stuff out.  Under the covers we use: https://github.com/jamesward/bower-as-a-service
For example:
https://bower-as-a-service.herokuapp.com/info/iron-a11y-announcer
https://bower-as-a-service.herokuapp.com/info/iron-a11y-announcer/1.0.5

Yeah, there definitely could be name collisions with the way we generate those artifact id's.  I'm not quite sure how to address that though...

**Comment by bartolom** [comment]
maybe something simple like two dashes
`github-com--spring-project--spring-framework`

`github-com--foo--bar-baz`
`github-com--foo-bar--baz`

**Comment by Legioth** [comment]
I would like to propose the opposite of the original suggestion: always expand everything to use an `artifactId` based on the full GitHub URL (e.g. `github-com-polymer-polymer`). The result would thus be that two different packages would have the same Maven dependency regardless of which form they have used in their `bower.json`, thus getting rid of duplicates on the classpath.

This might maybe be too radical since those short names are of course also handy. As a compromise, it could be considered to only apply the expansion logic for packages that are themselves generated with an `artifactId` based on the full URL. In that way, you could opt-in to getting consistent dependencies.

**Comment by jamesward** [comment]
I think the problem with always using the GitHub name is that in Bower there are at least two different ways you can specify deps, either on a Bower artifact or on a Git repo.  And these are technically different.  For instance, the Git repo might not have a version tag that was deployed to Bower.

**Comment by Legioth** [comment]
Isn't registration only about creating a mapping from a shorthand name to a Git URL? Once registered, versions tagged in Git are also available to anyone using the shorthand name.

See https://bower.io/docs/creating-packages/#register for a definition of what registration means. The resulting registry at https://registry.bower.io/packages only contains name mapping but no version numbers.

**Comment by jamesward** [comment]
Ah, so we can always go from an artifact name to a Git URL.  But can we go the other way cause it would be nice to actually transform everything (that has been registered) to an artifact name.

**Comment by Legioth** [comment]
Going the other way is more challenging since the registry can have multiple names mapped to the same Git URL. 

For instance `https://github.com/Polymer/polymer.git` has four different names:
* `Polymer`
* `polymer`
* `polymer-polymer`
* `GoogleWebComponents`

We could in theory look at the `name` field in `bower.json` to choose which one to use, even though there's no guarantee that the module name will be among the list of registered names.

**Comment by bartolom** [comment]
One related thing that popped up for us while trying to upgrade to Polymer 2.0 with webjars is:
> "The v prefix is allowed." From  https://bower.io/docs/creating-packages/#register

For Polymer 2.0 there is now a `v2.0.2` there instead of a `2.0.2` which is problematic for transitive dependencies in Gradle

I filed an issue with the polymer project https://github.com/Polymer/polymer/issues/4794 but from their JavaScript point of view this is all valid

**Comment by jamesward** [comment]
Fixed via #1702
